src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,20 +245,40 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
             }
         }
 
-        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer)
+        private unsafe void ParseEventBufferAndNotifyForEach(byte[] buffer, uint numBytes)
         {
             Debug.Assert(buffer != null);
             Debug.Assert(buffer.Length > 0);
 
+            numBytes = Math.Min(numBytes, (uint)buffer.Length);
+
             fixed (byte* b = buffer)
             {
+                byte* pBufferLimit = b + numBytes;
+
                 Interop.Kernel32.FILE_NOTIFY_INFORMATION* info = (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)b;
 
                 ReadOnlySpan<char> oldName = ReadOnlySpan<char>.Empty;
 
                 // Parse each event from the buffer and notify appropriate delegates
                 do
                 {
+                    // Validate the data we received in case it's corrupted. This can happen
+                    // if we are watching files over the network with a poor connection.
+
+                    // Verify the info object is within the expected bounds
+                    if (info < b || (((byte*)info) + sizeof(Interop.Kernel32.FILE_NOTIFY_INFORMATION)) > pBufferLimit)
+                    {
+                        break;
+                    }
+
+                    // Verify the file path is within the bounds
+                    byte* pFileEnd = ((byte*)&(info->FileName)) + info->FileNameLength;
```

Comment:
create a local for `info->FileName` and `info->FileNameLength` since they are also referenced below.

Reply:
To improve readability or to improve performance? It won't help to improve performance.

Reply:
To improve readability.

Reply:
I have submitted https://github.com/dotnet/runtime/pull/42474 that shows how this can be rewritten using Span<T> - to both improve reliability and improve security.

Reply:
@jkotas is the intention of that fix to be the backport for 3.1 and 5.0? In that case, I can close my 3 PRs and I can let you submit the backports as well.

Reply:
I think it is fine to use your original fix for servicing. It is less a bit less invasive.

Reply:
Okay great. So the only one I'll close is the one for master.

Reply:
On a second thought, if we want to preserve consistency, I should also merge the master change and your PR can go on top of that.

Reply:
Merging your PR to master sounds good to me.